lib/toolbox/strint.c

```diff
@@ -0,0 +1,121 @@
+ * the bounds of the integer.
+ * @param [in] str Input string
+ * @param [out] end Pointer to first character after the number in input string
+ * @param [out] abs_out Absolute part of result
+ * @param [out] sign_out Sign part of result (true=negative, false=positive)
+ * @param [in] base Integer base
+ * @param [in] max_abs_negative Largest permissible absolute part of result if
+ * the sign is negative
+ * @param [in] max_positive Largest permissible absolute part of result if the
+ * sign is positive
+ */
+StrintParseError strint_to_uint64_internal(
+    const char* str,
+    char** end,
+    uint64_t* abs_out,
+    bool* sign_out,
+    uint8_t base,
+    uint64_t max_abs_negative,
+    uint64_t max_positive) {
+    // skip whitespace
+    while(((*str >= 9) && (*str <= 13)) || *str == 32) {
+        str++;
+    }
+
+    // read sign
+    bool negative = false;
+    if(*str == '+' || *str == '-') {
+        if(*str == '-') negative = true;
+        str++;
+    }
+    if(*str == '+' || *str == '-') return StrintParseSignError;
+    if(max_abs_negative == 0 && negative) return StrintParseSignError;
+
+    // infer base
+    // not assigning directly to `base' to permit prefixes with explicit bases
+    uint8_t inferred_base = 0;
+    if(memcmp(str, "0x", 2) == 0 || memcmp(str, "0X", 2) == 0) {
+        inferred_base = 16;
+        str += 2;
+    } else if(memcmp(str, "0b", 2) == 0 || memcmp(str, "0B", 2) == 0) {
+        inferred_base = 2;
+        str += 2;
+    } else if(*str == '0') {
+        inferred_base = 8;
+        str++;
+    } else {
+        inferred_base = 10;
+    }
+    if(base == 0) base = inferred_base;
+
+    // read digits
+    uint64_t limit = negative ? max_abs_negative : max_positive;
+    uint64_t mul_limit = limit / base;
+    uint64_t result = 0;
+    int read_total = 0;
+    while(*str != 0) {
+        int digit_value;
+        if(*str >= '0' && *str <= '9') {
+            digit_value = *str - '0';
+        } else if(*str >= 'A' && *str <= 'Z') {
+            digit_value = *str - 'A' + 10;
+        } else if(*str >= 'a' && *str <= 'z') {
+            digit_value = *str - 'a' + 10;
+        } else {
+            break;
+        }
+
+        if(digit_value >= base) {
+            break;
+        }
+
+        if(result > mul_limit) return StrintParseOverflowError;
+        result *= base;
+        if(result > limit - digit_value) return StrintParseOverflowError;
+        result += digit_value;
+
+        read_total++;
+        str++;
+    }
+
+    if(read_total == 0) {
+        if(inferred_base == 8) {
+            // there's just a single zero
+            result = 0;
+        } else {
+            return StrintParseAbsentError;
+        }
+    }
+
+    if(abs_out) *abs_out = result;
+    if(sign_out) *sign_out = negative;
+    if(end) *end = (char*)str;
+    return StrintParseNoError;
+}
+
+#define STRINT_MONO(name, type, neg, pos)                                              \
```

Comment:
Please add more context for names, like `ret_type`, `neg_limit`, `pos_limit`, or write a comment